### PR TITLE
introduce ddt, test already-mapped circuits in the transpiler

### LINKED ISSUE #2532: Transpiler Errors
I created a very simple circuit

```
qr = qiskit.QuantumRegister(5)
cr = qiskit.ClassicalRegister(2)
qc = qiskit.QuantumCircuit(qr,cr)
qc.cx(2,4)
backend = qiskit.IBMQ.get_backend('ibmq_16_melbourne')
```
Importantly, `ibmq_16_melbourne` has no [2,4] connection.

If I transpile with `optimization_level=0` or `=1` I get the error in https://github.com/Qiskit/qiskit-terra/issues/2503, but if I try higher optimization levels I get

```
---------------------------------------------------------------------------
TranspilerError                           Traceback (most recent call last)
<ipython-input-18-363a9bd4512c> in <module>
----> 1 new_qv = qiskit.compiler.transpile(qv_circs[0][0], backend, optimization_level=2)

~/quantum_git/qiskit-sdk-py/qiskit/compiler/transpile.py in transpile(circuits, backend, basis_gates, coupling_map, backend_properties, initial_layout, seed_transpiler, optimization_level, pass_manager, seed_mapper)
    148 
    149     # Transpile circuits in parallel
--> 150     circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_configs)))
    151 
    152     if len(circuits) == 1:

~/quantum_git/qiskit-sdk-py/qiskit/tools/parallel.py in parallel_map(task, values, task_args, task_kwargs, num_processes)
     98     """
     99     if len(values) == 1:
--> 100         return [task(values[0], *task_args, **task_kwargs)]
    101 
    102     Publisher().publish("terra.parallel.start", len(values))

~/quantum_git/qiskit-sdk-py/qiskit/compiler/transpile.py in _transpile_circuit(circuit_config_tuple)
    169     circuit, transpile_config = circuit_config_tuple
    170 
--> 171     return transpile_circuit(circuit, transpile_config)
    172 
    173 

~/quantum_git/qiskit-sdk-py/qiskit/transpiler/transpile_circuit.py in transpile_circuit(circuit, transpile_config)
     60         pass_manager = default_pass_manager_simulator(transpile_config)
     61 
---> 62     return pass_manager.run(circuit)

~/quantum_git/qiskit-sdk-py/qiskit/transpiler/passmanager.py in run(***failed resolving arguments***)
    144         for passset in self.working_list:
    145             for pass_ in passset:
--> 146                 dag = self._do_pass(pass_, dag, passset.options)
    147 
    148         circuit = dag_to_circuit(dag)

~/quantum_git/qiskit-sdk-py/qiskit/transpiler/passmanager.py in _do_pass(self, pass_, dag, options)
    173             if pass_.is_transformation_pass:
    174                 pass_.property_set = self.fenced_property_set
--> 175                 new_dag = pass_.run(dag)
    176                 if not isinstance(new_dag, DAGCircuit):
    177                     raise TranspilerError("Transformation passes should return a transformed dag."

~/quantum_git/qiskit-sdk-py/qiskit/transpiler/passes/mapping/cx_direction.py in run(self, dag)
     79                 if self.coupling_map.distance(physical_q0, physical_q1) != 1:
     80                     raise TranspilerError('The circuit requires a connection between physical '
---> 81                                           'qubits %s and %s' % (physical_q0, physical_q1))
     82 
     83                 if (physical_q0, physical_q1) not in self.coupling_map.get_edges():

TranspilerError: 'The circuit requires a connection between physical qubits 4 and 2'
```

### Information

- **Qiskit Terra version**: 0.8.1 (stable tagged)
- **Python version**: 3.7
- **Operating system**: Mac

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #2532 

The problem with #2532 is that a layout selection makes the swapper unnecessary. However, running the swapper is not currently optional. Swapper flattens the quantum register space and the passes after it expect this transformation.

### CODE REVIEW

**Comment by ajavadia** [comment]
Can you add the failing example as a test? Also why is this a problem only for level 2, but not others?

Also can you open an issue to remember to restore the optional skipping of the swapper?
We will either:
- always flatten, in which case the swapper should not be responsible for this, but a pass before the swapper.
- never flatten
But this decision should be independent of this pipeline optimization where, if you don't need swaps, you should not run the swapper.

**Comment by 1ucian0** [comment]
extended to level 3

**Comment by 1ucian0** [comment]
In PR #2559 is being introduced. So, this test uses that.

File: qiskit/test/decorators.py [hunk 1]
```diff
@@ -204,3 +206,40 @@ def _wrapper(self, *args, **kwargs):
...
+                 name='{circuit.__name__}_level{level}')
+    """
+
+    def deco(func):
+        return data(*generate_cases(**kwargs))(unpack(func))
+
+    return deco
+
+
+METATEST = ddt
```
Comment (ajavadia) [inline qiskit/test/decorators.py]: I think using just `ddt` is more explicit about what this decorator is, as "metatest" does not have a very standard definition.

Comment (1ucian0) [reply qiskit/test/decorators.py]: fixed in 72307715

**Comment by ajavadia** [comment]
#2670 will fix the bug associated with this PR in a cleaner way. 

But the tests here are still useful, so I will undo the PassManager changes and keep the tests.

**Comment by 1ucian0** [comment]
#2670 was not merged. What's the next step?

**Comment by 1ucian0** [comment]
@ajavadia feel free to modify this branch to take tests from here

**Comment by 1ucian0** [comment]
Okey @ajavadia , have a look!

**Pull request merged**